java-checks/src/main/java/org/sonar/java/checks/ThreadLocalWithInitialCheck.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2017 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import org.sonar.check.Rule;
+import org.sonar.java.JavaVersionAwareVisitor;
+import org.sonar.java.checks.methods.AbstractMethodDetection;
+import org.sonar.java.matcher.MethodMatcher;
+import org.sonar.plugins.java.api.JavaVersion;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.NewClassTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+import java.util.Collections;
+import java.util.List;
+
+@Rule(key = "S4065")
+public class ThreadLocalWithInitialCheck extends AbstractMethodDetection implements JavaVersionAwareVisitor {
+  private static final MethodMatcher THREADLOCAL_CONSTRUCTOR = MethodMatcher.create().typeDefinition("java.lang.ThreadLocal").name("<init>").withoutParameter();
+
+  @Override
+  protected List<MethodMatcher> getMethodInvocationMatchers() {
+    return Collections.singletonList(THREADLOCAL_CONSTRUCTOR);
+  }
+
+  @Override
+  protected void onConstructorFound(NewClassTree newClassTree) {
+    ClassTree classTree = newClassTree.classBody();
+    if (classTree == null) {
+      return;
+    }
+    List<Tree> members = classTree.members();
+    if (members.size() != 1) {
+      return;
+    }
+    members.stream()
+      .filter(tree -> tree.is(Tree.Kind.METHOD))
+      .map(t -> ((MethodTree) t))
+      .filter(t -> "initialValue".equals(t.simpleName().name()))
```

Comment:
I know that it's probably I pretty nasty corner case, but as you are not checking the parameters, you may end up with FP when the `initialValue` method is not the override. Up to you if you want to fix it or not.

```
  ThreadLocal<List<String>> threadLocal = new ThreadLocal<List<String>>() {
      protected java.util.List<String> initialValue(String s) { // FP - not the override method but a different method
        List<String> init = initialValue();
        init.add(s);
        return init;
      };
    };
```